cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -96,73 +96,111 @@ predicate hasSize(HeuristicAllocationExpr alloc, DataFlow::Node n, int state) {
  */
-module SizeBarrier {
+private module SizeBarrier {
   private module SizeBarrierConfig implements DataFlow::ConfigSig {
     predicate isSource(DataFlow::Node source) {
       // The sources is the same as in the sources for the second
       // projection in the `AllocToInvalidPointerConfig` module.
       hasSize(_, source, _)
     }
 
+    /**
+     * Holds if `left <= right + k` holds if `g` evaluates to `testIsTrue`.
+     */
     additional predicate isSink(
       DataFlow::Node left, DataFlow::Node right, IRGuardCondition g, int k, boolean testIsTrue
     ) {
       // The sink is any "large" side of a relational comparison. i.e., the `right` expression
-      // in a guard such as `left < right + k`.
-      g.comparesLt(left.asOperand(), right.asOperand(), k, true, testIsTrue)
+      // in a guard such as `left <= right + k`.
+      g.comparesLt(left.asOperand(), right.asOperand(), k + 1, true, testIsTrue)
     }
 
     predicate isSink(DataFlow::Node sink) { isSink(_, sink, _, _, _) }
   }
 
   private import DataFlow::Global<SizeBarrierConfig>
 
-  private int getAFlowStateForNode(DataFlow::Node node) {
+  private int getASizeAddend(DataFlow::Node node) {
     exists(DataFlow::Node source |
       flow(source, node) and
       hasSize(_, source, result)
     )
   }
 
+  /**
+   * Holds if `left <= nRight + k` holds if `g` evaluates to `edge`.
+   */
   private predicate operandGuardChecks(
-    IRGuardCondition g, Operand left, Operand right, int state, boolean edge
+    IRGuardCondition g, Operand left, DataFlow::Node right, int k, boolean edge
   ) {
-    exists(DataFlow::Node nLeft, DataFlow::Node nRight, int k |
-      nRight.asOperand() = right and
-      nLeft.asOperand() = left and
-      SizeBarrierConfig::isSink(nLeft, nRight, g, k, edge) and
-      state = getAFlowStateForNode(nRight) and
-      k <= state
+    flowTo(right) and
+    SizeBarrierConfig::isSink(DataFlow::operandNode(left), right, g, k, edge)
+  }
+
+  /**
+   * Gets an instruction `instr` that is guarded by a check such as `instr <= left + delta` where
+   * `left <= _ + k` and `left` is the "small side" of of a relational comparison that checks
+   * whether `left <= size` where `size` is the size of an allocation.
+   */
+  Instruction getABarrierInstruction0(int delta, int k) {
+    exists(IRGuardCondition g, ValueNumber value, Operand left, boolean edge, DataFlow::Node right |
+      left = value.getAUse() and
+      // We know:
+      // 1. result <= value + delta (by `bounded`)
+      // 2. value <= right + k (by `operandGuardChecks`).
+      // So:
+      // result <= value + delta (by 1.)
+      //        <= right + k + delta (by 2.)
+      operandGuardChecks(pragma[only_bind_into](g), pragma[only_bind_into](left), right,
+        pragma[only_bind_into](k), pragma[only_bind_into](edge)) and
+      bounded(result, value.getAnInstruction(), delta) and
+      g.controls(result.getBlock(), edge) and
+      k < getASizeAddend(right)
```

Comment:
I'm still unsure why this restriction is needed, or what is does exactly.

Reply:
This ensures that the `k` we infer in a guard such as `myIndex < size + k` isn't "too large". For example, the guard `n < size + 999` should not be used to determine that the pointer-arithmetic `p + n` creates a pointer that's within bounds:
```cpp
int* p = new[size];
if(n < size + 999) {
  int* end = p + n;
}
```
in this case `k` would be `999`, and `getASizeAddend(right)` would be `0` (because we allocated `size + 0` elements).

Does that make sense?

Reply:
That seems to suggest that we should rule out any `k> 0`, or not?

Reply:
I think that's too strict. For example, consider the case where the first line is:
```
int* p = new[size + 1];
```
then a condition such as `if(n < size + 1)` should be a barrier.

Reply:
Maybe I'm confused again, even after reading the relevant code again. I though that the size expression that we're tracking in this case would be `size + 1` and that this expression is what we're ultimately also tracking in the dataflow associated with this barrier.

Reply:
We track `size` in both the "product-flow configuration" and in the "barrier configuration". That's because it's not possible to track the value `size + 1` (since there's no flow out of this expression). So in the product-flow configuration we "remember" that, while we're tracking `size`, the actual size of the allocation is `size + 1` by keeping a flow state with the value `1`. In a program like:
```cpp
1. int* p = new int[size + 1];
2. if(x < size) {
3.   ...
4. }
```
The result of `getASizeAddend(right)` (where `right` is the dataflow node for `size` on line 2) would give us `1` since `size` is used in an allocation that has a "size addend" of `2`.

Reply:
So I'm confused apparently.

Reply:
Let me know if a face-to-face discussion would help!

Reply:
Right, I was apparently not able to read the `hasSize` implementation correctly. Looking at it again, what you say makes sense.